cache_images/local-cache-registry

```diff
@@ -219,6 +220,7 @@ function cache_image() {
     for retry in 1 2 3 0;do
         skopeo --registries-conf /dev/null \
                copy --all --dest-tls-verify=false \
+               --dest-compress-format zstd:chunked \
```

Comment:
This changes image digests, which are hard-coded in tests (e.g. `ALPINELISTDIGEST`).

That’s fine in principle — and I’ll update the test digests in the integration PR — I’m just noting that it’s expected that some tests will start failing with this.

---

Outside of the current zstd:chunked testing effort, consider using `--preserve-digests` in the command to make any accidental digest changes (e.g. if an unusual image weren’t supported by the cache registry and triggered a format conversion) an error, to make sure this creates a bit-for-bit-identical copy.

Reply:
Thanks for this. The new images built just now use `--preserve-digests`. Would you care to try with `c20241022t120000z-f40f39d13` ?

Reply:
I’m sorry, with `--preserve-digest` + `--dest-compress-format`, `--dest-compress-format` is, AFAICS, ignored (because that requires a digest change). I’ll inspect the images to confirm.

Arguably it’s a bug that this option combination is not rejected.

Reply:
Ack, thanks, I misunderstood.

So, back to your initial comment: "This changes image digests". Are these changes reliably consistent? When we make new VM images next week/year, different version of skopeo, different kernel/filesystem/tools, different phase of the moon, will the image digests be the same?

If so, then it might be okay to change the digests in the tests, **but** it's important to remember that e2e tests need to be able to run using local cache and also run using quay. Tests can determine the environment via `UsingCacheRegistry()` (returns bool). This is ugly but it's a potential way to get tests to pass.

If digests are not consistently reproducible, all bets are off and the tests will need to be completely reimagined.

Reply:
> Are these changes reliably consistent?

No; there is no guarantee that compression produces consistent results even with one version, and the output can (and, from time to time, does) change when the implementation changes.

For the purpose of the current experiment, I’ll hard-code the digests as they exist in the local cache. But, other than that, we’ll indeed need to specifically maintain zstd:chunked images on quay.io — or, alternatively, not hard-code the expected digests in tests, and obtain them at runtime, or maybe at the cache build time, using something like Skopeo.

Reply:
I think the digest always depends on the compression, so as long as the compression is reproducible it should produce the same digest?

In any case is there any need to hard code the digests? Skopeo inspect can give us that
```
skopeo inspect --no-tags  --format {{.Digest}} docker://quay.io/libpod/testimage:20241011
```
(The command seems quite slow for some reason, so it isn't something I would like to run in every test but only once and cache it)

Reply:
Compression isn’t reproducible.

Using `skopeo inspect` is definitely an option — noting that it decreases the test coverage a bit: if c/image were broken, we’d see consistent incorrect data from both Skopeo and the Podman binary being tested.

Reply:
Sure — and the risk is not that high.

---

Maybe, first, we need some thought about the final test plan: do we actually want to add a gzip / zstd:chunked _dimension_ to the testing matrix, to make this a variable? Maybe the necessary testing coverage can be achieved by a few dedicated tests/images. That’s very unclear to me.

Reply:
Exactly. This is one of the questions on my radar, that I don't yet know enough to answer. My hope is that we can do an integration, mixed set of images, but as of this writing my recollection is that some images have worked (with zstd:chunked) and some have not, and that makes me uncomfortable. It's possible that the reason for this is that some _tests_ use different images and that's what triggers the failures, but I haven't spent the time to investigate. Thank you for bringing this up.